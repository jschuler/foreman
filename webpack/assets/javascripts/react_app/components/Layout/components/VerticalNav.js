import React from 'react';
import {
  Nav,
  NavExpandable,
  NavItem,
  NavItemSeparator,
  NavList,
} from '@patternfly/react-core';
import {
  getCurrentPath,
  handleMenuClick,
  layoutPropTypes,
  layoutDefaultProps,
} from '../LayoutHelper';

// checkPropTypes will throw a warning:
// Warning: Failed prop type: Invalid prop `title` of type `object` supplied to `NavExpandable`, expected `string`.
// This is expected until PatternFly is updated to accept ReactNode instead of just string for the title
const titleWithIcon = (title, iconClass) => (
  <React.Fragment>
    <span
      className={iconClass}
      style={{ pointerEvents: 'none', paddingRight: '10px' }}
    />
    <span style={{ pointerEvents: 'none' }}>{title}</span>
  </React.Fragment>
);

const VerticalNav = ({
  items,
  activeMenu,
  changeActiveMenu,
  history,
  ...props
}) => {
  const pathFragment = path => path.split('?')[0];

  // the pages outside of React don't have activeMenu information.
  // workaround is to build our own mapping that we can reference for the nav
  const subItemToItemMap = {};
  items.forEach(item => {
    item.subItems.forEach(subItem => {
      if (!subItem.isDivider) {
        // don't keep the query parameters for the key
        subItemToItemMap[pathFragment(subItem.href)] = item.title;
      }
    });
  });

  return (
    <Nav theme="dark" {...props}>
      <NavList>
        {items.map(item => (
          <NavExpandable
            key={item.title}
            title={titleWithIcon(item.title, item.iconClass)}
            isActive={
              subItemToItemMap[pathFragment(getCurrentPath())] === item.title
            }
            isExpanded={
              subItemToItemMap[pathFragment(getCurrentPath())] === item.title
            }
            className="foreman-nav-expandable"
          >
            {item.subItems.map(subItem =>
              subItem.isDivider ? (
                <NavItemSeparator key={subItem.id} />
              ) : (
                <NavItem
                  key={subItem.id || subItem.href}
                  preventDefault
                  onClick={() => {
                    handleMenuClick(item, activeMenu, changeActiveMenu);
                    history.push(subItem.href);
                  }}
                  isActive={
                    pathFragment(subItem.href) ===
                    pathFragment(getCurrentPath())
                  }
                >
                  {subItem.title}
                </NavItem>
              )
            )}
          </NavExpandable>
        ))}
      </NavList>
    </Nav>
  );
};
VerticalNav.propTypes = {
  activeMenu: layoutPropTypes.activeMenu,
  changeActiveMenu: layoutPropTypes.changeActiveMenu,
  history: layoutPropTypes.history.isRequired,
  items: layoutPropTypes.items,
};
VerticalNav.defaultProps = {
  activeMenu: layoutDefaultProps.activeMenu,
  changeActiveMenu: layoutDefaultProps.changeActiveMenu,
  items: layoutDefaultProps.items,
};
export default VerticalNav;
