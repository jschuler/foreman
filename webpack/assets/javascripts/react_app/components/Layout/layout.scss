@import '../../common/colors.scss';

.pf-c-page.foreman-theme {
  .pf-c-page__header {
    .pf-c-data-toolbar {
      background-color: transparent;
    }
  }

  .pf-c-page__sidebar.pf-m-dark {
    background-color: $nav-pf-vertical-secondary-active-bg-color;

    .pf-c-nav.pf-m-dark > .pf-c-nav__list {
      .pf-c-nav__item.pf-m-current,
      .pf-c-nav__item > .pf-c-nav__link.pf-m-current {
        background-color: $nav-pf-vertical-active-bg-color;
      }

      .pf-c-nav__item.pf-m-current > .pf-c-nav__link[data-component=pf-nav-expandable] {
        color: $nav-pf-vertical-active-icon-color;
      }

      .pf-c-nav__link {
        color: $nav-pf-vertical-icon-color;
      }

      .pf-c-nav__simple-list .pf-c-nav__link {
        &.pf-m-current,
        &.pf-m-hover,
        &:hover {
          background-color: $nav-pf-vertical-active-bg-color;
          color: $nav-pf-vertical-active-icon-color;
        }
      }
    }
  }
}

#account_menu {
  outline: none;
}

#vertical-spinner {
  margin-top: 15px;
  margin-left: 5px;
}
