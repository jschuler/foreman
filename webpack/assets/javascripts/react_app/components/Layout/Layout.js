import React from 'react';
import {
  Brand,
  Page,
  PageSection,
  PageSectionVariants,
  PageSidebar,
  SkipToContent,
} from '@patternfly/react-core';
// eslint-disable-next-line
import globalBreakpointMd from '@patternfly/react-tokens/dist/js/global_breakpoint_md';
import CustomPageHeader from './components/CustomPageHeader';
import {
  getActive,
  getCurrentPath,
  layoutPropTypes,
  layoutDefaultProps,
} from './LayoutHelper';
import VerticalNav from './components/VerticalNav';
import Toolbar from './components/Toolbar';
import './layout.scss';

class Layout extends React.Component {
  componentDidMount() {
    const {
      items,
      data,
      fetchMenuItems,
      changeLocation,
      currentLocation,
      changeOrganization,
      currentOrganization,
      changeActiveMenu,
      activeMenu,
    } = this.props;
    if (items.length === 0) fetchMenuItems(data);

    const activeURLMenu = getActive(data.menu, getCurrentPath());
    if (activeMenu !== activeURLMenu.title) {
      changeActiveMenu(activeURLMenu);
    }

    if (
      data.taxonomies.locations &&
      !!data.locations.current_location &&
      currentLocation !== data.locations.current_location
    ) {
      const initialLocTitle = data.locations.current_location;
      const initialLocId = data.locations.available_locations.find(
        loc => loc.title === initialLocTitle
      ).id;
      changeLocation({ title: initialLocTitle, id: initialLocId });
    }

    if (
      data.taxonomies.organizations &&
      !!data.orgs.current_org &&
      currentOrganization !== data.orgs.current_org
    ) {
      const initialOrgTitle = data.orgs.current_org;
      const initialOrgId = data.orgs.available_organizations.find(
        org => org.title === initialOrgTitle
      ).id;
      changeOrganization({ title: initialOrgTitle, id: initialOrgId });
    }
  }

  render() {
    const {
      items,
      data,
      isLoading,
      changeActiveMenu,
      changeOrganization,
      changeLocation,
      currentOrganization,
      currentLocation,
      activeMenu,
      children,
      history,
    } = this.props;

    const afterNavToggle = navOpen => {
      const mobileView =
        window.innerWidth < Number.parseInt(globalBreakpointMd.value, 10);
      if (mobileView) {
        // ignore for mobile view
        return;
      }
      // toggles a class in the body tag, so that the main #rails-app-content container can have the appropriate width
      if (navOpen) {
        document.body.classList.add('pf-m-expanded');
      } else {
        document.body.classList.remove('pf-m-expanded');
      }
    };
    const header = (
      <CustomPageHeader
        logo={
          <Brand
            src={data.logo}
            alt={data.brand}
            style={{ marginTop: '8px' }}
          />
        }
        logoProps={{ href: data.root }}
        showNavToggle
        contextSelector={
          <Toolbar
            data={data}
            currentLocation={currentLocation}
            changeLocation={changeLocation}
            currentOrganization={currentOrganization}
            changeOrganization={changeOrganization}
            isLoading={isLoading}
            changeActiveMenu={changeActiveMenu}
          />
        }
        afterNavToggle={afterNavToggle}
        className="navbar-pf-vertical"
      />
    );
    const sidebar = (
      <PageSidebar
        nav={
          <VerticalNav
            items={items}
            activeMenu={activeMenu}
            changeActiveMenu={changeActiveMenu}
            history={history}
          />
        }
        theme="dark"
      />
    );
    const pageId = 'main';
    const pageSkipToContent = (
      <SkipToContent href={`#${pageId}`}>Skip to content</SkipToContent>
    );

    return (
      <Page
        header={header}
        sidebar={sidebar}
        isManagedSidebar
        skipToContent={pageSkipToContent}
        mainContainerId={pageId}
        className="foreman-theme"
      >
        <PageSection variant={PageSectionVariants.light}>
          {children}
        </PageSection>
      </Page>
    );
  }
}

Layout.propTypes = {
  ...layoutPropTypes,
};

Layout.defaultProps = {
  ...layoutDefaultProps,
};

export default Layout;
