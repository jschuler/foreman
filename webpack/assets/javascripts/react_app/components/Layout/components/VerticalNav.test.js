import { testComponentSnapshotsWithFixtures } from '../../../common/testHelpers';
import { hasTaxonomiesMock } from '../Layout.fixtures';

import VerticalNav from './VerticalNav';

const fixtures = {
  'render VerticalNav': {
    activeMenu: hasTaxonomiesMock.activeMenu,
    changeActiveMenu: () => {},
    history: hasTaxonomiesMock.history,
    items: hasTaxonomiesMock.items,
  },
};

describe('VerticalNav', () => {
  describe('rendering', () => {
    const spy = jest.spyOn(console, 'error').mockImplementation();

    testComponentSnapshotsWithFixtures(VerticalNav, fixtures);

    // Currently expect a prop warning since VerticalNav passes an object into NavExpandable title prop although it expects a string. This will change soon.
    expect(spy).toHaveBeenCalledTimes(1);
    // eslint-disable-next-line
    expect(console.error.mock.calls[0][0]).toEqual(expect.stringContaining('Warning: Failed prop type: Invalid prop `title` of type `object` supplied to `NavExpandable`, expected `string`'));
    spy.mockRestore();
  });
});
