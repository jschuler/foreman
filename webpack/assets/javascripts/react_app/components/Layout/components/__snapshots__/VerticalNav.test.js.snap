// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VerticalNav rendering render VerticalNav 1`] = `
<Component
  theme="dark"
>
  <NavList
    ariaLeftScroll="Scroll left"
    ariaRightScroll="Scroll right"
    className=""
    variant="default"
  >
    <NavExpandable
      className="foreman-nav-expandable"
      groupId={null}
      id=""
      isActive={false}
      isExpanded={false}
      key="Monitor"
      srText=""
      title={
        <React.Fragment>
          <span
            className="fa fa-tachometer"
            style={
              Object {
                "paddingRight": "10px",
                "pointerEvents": "none",
              }
            }
          />
          <span
            style={
              Object {
                "pointerEvents": "none",
              }
            }
          >
            Monitor
          </span>
        </React.Fragment>
      }
    >
      <NavItem
        isActive={false}
        key="/Aa"
        onClick={[Function]}
        preventDefault={true}
      >
        Aa
      </NavItem>
      <NavItem
        isActive={false}
        key="/Cc"
        onClick={[Function]}
        preventDefault={true}
      >
        Cc
      </NavItem>
    </NavExpandable>
    <NavExpandable
      className="foreman-nav-expandable"
      groupId={null}
      id=""
      isActive={false}
      isExpanded={false}
      key="Hosts"
      srText=""
      title={
        <React.Fragment>
          <span
            className="fa fa-server"
            style={
              Object {
                "paddingRight": "10px",
                "pointerEvents": "none",
              }
            }
          />
          <span
            style={
              Object {
                "pointerEvents": "none",
              }
            }
          >
            Hosts
          </span>
        </React.Fragment>
      }
    >
      <NavItem
        isActive={false}
        key="/Dd"
        onClick={[Function]}
        preventDefault={true}
      >
        Dd
      </NavItem>
      <NavItem
        isActive={false}
        key="/Ff"
        onClick={[Function]}
        preventDefault={true}
      >
        Ff
      </NavItem>
    </NavExpandable>
    <NavExpandable
      className="foreman-nav-expandable"
      groupId={null}
      id=""
      isActive={false}
      isExpanded={false}
      key="Configure"
      srText=""
      title={
        <React.Fragment>
          <span
            className="fa fa-wrench"
            style={
              Object {
                "paddingRight": "10px",
                "pointerEvents": "none",
              }
            }
          />
          <span
            style={
              Object {
                "pointerEvents": "none",
              }
            }
          >
            Configure
          </span>
        </React.Fragment>
      }
    >
      <NavItem
        isActive={false}
        key="/Gg"
        onClick={[Function]}
        preventDefault={true}
      >
        Gg
      </NavItem>
      <NavItem
        isActive={false}
        key="/Ii"
        onClick={[Function]}
        preventDefault={true}
      >
        Ii
      </NavItem>
    </NavExpandable>
  </NavList>
</Component>
`;
